### ISSUE TITLE
Similarly named typedefs in adjacent modules cause overlap.

### ISSUE THREAD

**Issue by EliteMasterEric** [HaxeFoundation/haxe#11079]
Assume the contents of module `A.hx`:
```haxe
class A {
    public function new() {}
}

class A_Builder {
    public static function build():A { return new A(); }
}

typedef Builder = A_Builder;
```

...and of module `B.hx`:
```haxe
class B {
    public function new() {}
}

class B_Builder {
    public static function build():B { return new B(); }
}

typedef Builder = B_Builder;
```

As long as I am not defining a typedef that is the same name as an existing module, my expectation is that I would like to be able to refer to `com.example.A.Builder` as `A.Builder` elsewhere in my code, but I get an error that `Builder` is already defined in this package.

My use case is that I am creating a set of externs for Java, and I have a set of classes which have inner classes called `Builder`. The first extern I create works as expected, but the second extern I create fails, and I am forced to prefix the name of the extern to distinguish them to Haxe. This prefixing forces me to change the function name, which means the Haxe library's names are mangled compared to the underlying API.

**Comment by Simn** [comment]
This annoys me sometimes too, but it's how module types work in Haxe. I think the best solution would be something that behaves like `private typedef` regarding its path, but can still be accessed like a non-private typedef from outside the module. Not sure how to express this in syntax though.

**Comment by EliteMasterEric** [comment]
Agreed, I can keep the actual underlying class names unique (since they're just externs) but I would definitely at least like the users to be able to access them using the intended inner class name.

**Comment by EliteMasterEric** [comment]
I would still like to see this issue resolved. As of right now, I just remove the typedefs (but only when they cause incompatibilities). So still causing a little bit of mangling of names for subclasses but not all the names.

**Comment by Simn** [comment]
Well, I'm open to ideas regarding the syntax here.

**Comment by EliteMasterEric** [comment]
Not sure what the best solution to this is. Maybe introducing the `protected` keyword, so `protected typedef`? Introducing a whole new reserved word for this niche scenario seems weird but `protected` can be reused for other use cases and has precedence as a reserved word in languages like Java.

**Comment by kLabz** [comment]
I don't really like `protected`, especially for this use case. Maybe `local` would convey some of the meaning here? The type is not exposed to the whole module by default, but not being "private" it can be accessed by other mean (full path) when needed.

**Comment by EliteMasterEric** [comment]
To me, `local` implies that the `typedef` is only accessible within the same module it was defined in, like `private typedef` (which I JUST NOW realized it a thing, learn something new every day).

**Comment by Simn** [comment]
The confusion here is because there are technically two kinds of `private` for types:

1. The output one which makes the path unique by using `package.Module._PrivateType` paths.
2. The access one which disallows accessing these types from other modules.

The feature request is to have 1. without 2.

**Comment by Simn** [comment]
Maybe we could go for a simple `@:allow private typedef` because that works for fields and such as well.

**Comment by kLabz** [comment]
Would @:privateAccess work too ?

**Comment by Simn** [comment]
Yes that should be supported too. I think there's an old open issue about this problem related to macros, where you can't always access a type by its path because it could be private.

**Comment by back2dos** [comment]
Well, being able to refer to private types directly would be nice, but I think it's a different issue.

The main problem is that for some reason `some.pack.Module.ThatType` becomes `some.pack.ThatType` and I have two questions:

1. Does that really makes sense? What are even the advantages?
2. What would actually break if we said `some.pack.Module.ThatType` stays `some.pack.Module.ThatType` (or maybe with `_Module` to be in line with private types)? I mean, it's not even possible to refer to it as  `some.pack.Module.ThatType` as far as I can tell: https://try.haxe.org/#77465190. The one thing I can think of is macros expecting `ThatType` to match `{ pack: ['some', 'pack'], name: 'ThatType' }`, but if breaking those is all there is to it, that's a superb trade-off.

This would still make it a little awkward to refer to `some.pack.SomeModule.ThatType` and `some.pack.OtherModule.ThatType` in the same module, but at least the types could coexist in one build in the first place.

**Comment by EliteMasterEric** [comment]
> This would still make it a little awkward to refer to `some.pack.SomeModule.ThatType` and `some.pack.OtherModule.ThatType` in the same module, but at least the types could coexist in one build in the first place.

Can't you use simply use [import with alias](https://haxe.org/manual/type-system-import.html#import-with-alias)? `import some.pack.SomeModule.ThatType as Foo;` and `import some.pack.OtherModule.ThatType as Bar`?

**Comment by back2dos** [comment]
I said "a little awkward" and not "impossible". The awkwardness perhaps becomes more apparent when you consider two modules with the same structure where you want to import five types from each. But yes, having that problem is definitely an improvement over a failing build ;)

**Comment by Simn** [comment]
I understand the sentiment and this often annoys me too, but I doubt this could be changed by now without breaking many things. I also don't think that all our targets support something like that.

Allowing access to private types is a goal that can be achieved much easier, so I'd prefer to focus on that. If somebody wants to create a haxe-evolution for this kind of type path change I'm happy to discuss it.

**Comment by back2dos** [comment]
> I understand the sentiment and this often annoys me too, but I doubt this could be changed by now without breaking many things. I also don't think that all our targets support something like that.

I appreciate your sympathies, but I struggle to understand the issue. Could you please elaborate? Otherwise any haxe-evolution proposal I'd make would probably be rather useless.

> I also don't think that all our targets support something like that.

Welp, dunno. Let's take what looks like the simplest option to me: generate public sub types the same as private sub types. We know that that's supported. The question is what other issues might spring from that. I can think of:

1. The above mentioned macro issue.
2. Type.getClassName/Type.resolveClass (and enum counterparts) are affected. Not a problem per se, but haxe serialization of sub types would lead to different results depending on Haxe version.

Neither of these seem to be in the "we can never ever change this" category, so I wonder whether I'm missing some much bigger problem.

> Allowing access to private types is a goal that can be achieved much easier, so I'd prefer to focus on that. 

A valid point, although I'm not sure we really want `@:allow private class ...` to become a pattern.
